crates/wit-component/src/linking.rs

```diff
@@ -0,0 +1,1383 @@
+                Some((export.name, (export, exporters)))
+            } else {
+                None
+            }
+        })
+        .collect::<HashMap<_, _>>();
+
+    let mut resolved = HashMap::new();
+    let mut missing = Vec::new();
+    let mut duplicates = Vec::new();
+
+    let mut triage = |metadata: &'a Metadata, export: Export<'a>| {
+        if let Some((key, value)) = exporters.get_key_value(&export.key) {
+            match value.as_slice() {
+                [] => unreachable!(),
+                [exporter] => {
+                    resolved.insert(*key, *exporter);
+                }
+                _ => {
+                    duplicates.push((metadata.name, *key, value.as_slice()));
+                }
+            }
+        } else {
+            missing.push((metadata.name, export));
+        }
+    };
+
+    for metadata in metadata {
+        for (name, (ty, flags)) in &metadata.env_imports {
+            triage(
+                metadata,
+                Export {
+                    key: ExportKey {
+                        name,
+                        ty: Type::Function(ty.clone()),
+                    },
+                    flags: *flags,
+                },
+            );
+        }
+
+        for name in &metadata.memory_address_imports {
+            triage(
+                metadata,
+                Export {
+                    key: ExportKey {
+                        name,
+                        ty: Type::Global(GlobalType {
+                            ty: ValueType::I32,
+                            mutable: false,
+                        }),
+                    },
+                    flags: 0,
+                },
+            );
+        }
+    }
+
+    for metadata in metadata {
+        for name in &metadata.table_address_imports {
+            if let Some((key, value)) = function_exporters.get(name) {
+                match value.as_slice() {
+                    [] => unreachable!(),
+                    [exporter] => {
+                        resolved.insert(key, *exporter);
+                    }
+                    _ => {
+                        duplicates.push((metadata.name, *key, value.as_slice()));
+                    }
+                }
+            } else {
+                missing.push((
+                    metadata.name,
+                    Export {
+                        key: ExportKey {
+                            name,
+                            ty: Type::Function(FunctionType {
+                                parameters: Vec::new(),
+                                results: Vec::new(),
+                            }),
+                        },
+                        flags: 0,
+                    },
+                ));
+            }
+        }
+    }
+
+    (resolved, missing, duplicates)
+}
+
+/// Recursively add a library (represented by its offset) and its dependency to the specified set, maintaining
+/// topological order (modulo cycles).
```

Comment:
I'm a bit curious about how this works in the presence of cycles. I've never worked with topo sorts with cycles before, so naively I'd interpret the checks below as they handle a cycle of A->B->A but wouldn't handle deeper cycles and may recurse infinitely? I haven't worked with this before so I'm not sure though.

Put another way I suppose, what's the meaning of a topo sort in the presence of cycles?

Reply:
In this case, "topo sort in the presence of cycles" means the algorithm makes no guarantees about the relative order of items which participate in a cycle.  If we wanted to, we could refine that a bit so that if A needs 20 things from B and B only needs 3 things from A, then we'd put B first since it doesn't depend on A "as much" as A depends on B, but that was more than I was ready to tackle.

It handles deeper cycles by expecting that `dependencies` includes transitive dependencies, not just the first level of dependencies.